File: Baekjoon/BFS/BFS13549.cpp
#include <iostream>
#include <queue>
using namespace std;

//[숨바꼭질 3 난이도★★](https://www.acmicpc.net/problem/13549)

/*
	엄한데서 해맨 문제...
*/

int n, k;
int chk[100001];

int main() {
	cin >> n >> k;
	for (int i = 0; i < 100001; i++)
		chk[i] = 999999999;
	chk[n] = 0;
	queue<int> q;
	q.push(n);
	while (!q.empty()) {
		int now = q.front();
		q.pop();
		if (now == k) {
			cout << chk[now] << "\n";
			break;
		}
		//현재에서 -1 , 1 을한다.
		if (now + 1 <= 100000 && chk[now + 1] > chk[now] + 1) {
			q.push(now+1);
			chk[now + 1] = chk[now] + 1;
		}
		if (now - 1 >= 0 && chk[now - 1] > chk[now] + 1) {
			q.push(now - 1);
			chk[now - 1] = chk[now] + 1;
		}
		//2배를 한다.
		if (now * 2 < 100001 && chk[now * 2] > chk[now]) {
			q.push(now*2);
			chk[now * 2] = chk[now] ;
		}
	}
}

File: Baekjoon/BFS/BFS13549.kt
import java.util.*
//[숨바꼭질3](https://www.acmicpc.net/problem/13549)
fun main() = with(Scanner(System.`in`)){
    var (N,K) = nextLine().split(" ").map{it.toInt()}
    var chk = MutableList(100001){100001}
    var q : Queue<Int> = LinkedList()
    q.add(N)
    chk[N] = 0
    while(!q.isEmpty()){
        var pos = q.poll()
        if(chk[K] <= chk[pos]) continue
        if(pos > 0 && chk[pos-1] > chk[pos]+1) {
            chk[pos-1] = chk[pos]+1
            q.add(pos - 1)
        }
        if(pos < 100000 && chk[pos+1] > chk[pos]+1) {
            chk[pos+1] = chk[pos]+1
            q.add(pos + 1)
        }
        if(pos*2 <= 100000 && chk[pos*2] > chk[pos]) {
            chk[pos*2] = chk[pos]
            q.add(pos * 2)
        }
    }
    println(chk[K])
}